dhcpv4/client.go

```diff
@@ -165,14 +207,110 @@ func (c *Client) Exchange(ifname string, discover *DHCPv4, modifiers ...Modifier
+// response up to some read timeout value. If the message type is not
+// MessageTypeNone, it will wait for a specific message type
+func (c *Client) SendReceive(sendFd, recvFd int, packet *DHCPv4, messageType MessageType) (*DHCPv4, error) {
+	var laddr net.UDPAddr
+	if c.LocalAddr == nil {
+		laddr = net.UDPAddr{IP: net.IPv4bcast, Port: ServerPort}
+	} else {
+		if addr, ok := c.LocalAddr.(*net.UDPAddr); ok {
+			laddr = *addr
+		} else {
+			return nil, fmt.Errorf("Invalid local address: not a net.UDPAddr: %v", c.LocalAddr)
+		}
+	}
+	var raddr net.UDPAddr
+	if c.RemoteAddr == nil {
+		raddr = net.UDPAddr{IP: net.IPv4bcast, Port: ServerPort}
+	} else {
+		if addr, ok := c.RemoteAddr.(*net.UDPAddr); ok {
+			raddr = *addr
+		} else {
+			return nil, fmt.Errorf("Invalid remote address: not a net.UDPAddr: %v", c.RemoteAddr)
+		}
+	}
+	packetBytes, err := MakeRawPacket(packet.ToBytes(), &raddr, &laddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a goroutine to perform the blocking send, and time it out after
+	// a certain amount of time.
+	var (
+		destination [4]byte
+		response    *DHCPv4
+	)
+	copy(destination[:], raddr.IP.To4())
+	remoteAddr := unix.SockaddrInet4{Port: laddr.Port, Addr: destination}
+	recvErrors := make(chan error, 1)
+	go func(errs chan<- error) {
+		conn, innerErr := net.FileConn(os.NewFile(uintptr(recvFd), ""))
+		if err != nil {
+			errs <- innerErr
+			return
+		}
+		defer conn.Close()
+		conn.SetReadDeadline(time.Now().Add(c.ReadTimeout))
+
+		for {
+			buf := make([]byte, MaxUDPReceivedPacketSize)
+			n, _, _, _, innerErr := conn.(*net.UDPConn).ReadMsgUDP(buf, []byte{})
+			if innerErr != nil {
+				errs <- innerErr
+				return
+			}
+
+			response, innerErr = FromBytes(buf[:n])
+			if err != nil {
+				errs <- innerErr
+				return
+			}
+			// check that this is a response to our message
+			if response.TransactionID() != packet.TransactionID() {
+				continue
+			}
+			// wait for a response message
+			if response.Opcode() != OpcodeBootReply {
+				continue
+			}
+			// if we are not requested to wait for a specific message type,
+			// return what we have
+			if messageType == MessageTypeNone {
+				break
+			}
+			// break if it's a reply of the desired type, continue otherwise
+			if response.MessageType() != nil && *response.MessageType() == messageType {
+				break
+			}
+		}
+		recvErrors <- nil
+	}(recvErrors)
+	if err = unix.Sendto(sendFd, packetBytes, 0, &remoteAddr); err != nil {
+		return nil, err
+	}
+
+	select {
+	case err = <-recvErrors:
+		if err != nil {
+			return nil, err
+		}
+	case <-time.After(c.ReadTimeout):
+		return nil, errors.New("timed out while listening for replies")
+	}
+
+	return response, nil
+}
+
 // BroadcastSendReceive broadcasts packet (with some write timeout) and waits for a
 // response up to some read timeout value. If the message type is not
 // MessageTypeNone, it will wait for a specific message type
```

Comment:
Is BroadcastSendReceive being used? If not I would remove it.

Reply:
I will add a log to mark it as deprecated, something like 

```
Warning: dhcpv4.BroadcastSendAndReceive() is deprecated and will be removed. You should use dhcpv4.client.Exchange() instead.
```